### PR TITLE
Feature/context

### PR DESCRIPTION
**Pull request by sveisvei**
Closes https://github.com/openzipkin/zipkin-js/issues/2

### CODE REVIEW

File: packages/zipkin-context-cls/package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "zipkin-context-cls",
+  "version": "0.1.9",
+  "description": "A Context API implementation that uses continuation-local-storage under the hood",
+  "main": "src/CLSContext.js",
+  "scripts": {
+    "test": "node ../../node_modules/mocha/bin/mocha --require ../../test/helper.js"
+  },
+  "author": "Openzipkin <openzipkin.alt@gmail.com>",
```
Comment (adriancole) [inline packages/zipkin-context-cls/package.json]: kill me for mentioning the case format is OpenZipkin? :P

File: packages/zipkin/src/batch-recorder.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  _writeSpan(id) {
+    const spanToWrite = this.partialSpans.get(id);
+    // ready for garbage collection
+    this.partialSpans.delete(id);
+    this.logger.logSpan(spanToWrite);
+  }
+
+  _updateSpanMap(id, updater) {
```
Comment (adriancole) [inline packages/zipkin/src/batch-recorder.js]: heh looks like finagle!

File: packages/zipkin/src/batch-recorder.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+          this._annotate(span, rec, thriftTypes.SERVER_RECV);
+          break;
+        case 'Message':
+          this._annotate(span, rec, rec.annotation.message);
+          break;
+        case 'Rpc':
+          span.setName(rec.annotation.name);
+          break;
+        case 'ServiceName':
+          span.setServiceName(rec.annotation.serviceName);
```
Comment (adriancole) [inline packages/zipkin/src/batch-recorder.js]: generally, this is held constant per app, but I don't know the scope of this (and also recognise this is emulating finagle's behavior)

File: packages/zipkin/src/tracer/record.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Record {
```
Comment (adriancole) [inline packages/zipkin/src/tracer/record.js]: for types that are basically the same as finagle, should we link? or cover in overall readme that this is largely inspired by finagle's Tracer

File: packages/zipkin-context-cls/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+## Usage:
+
+```javascript
+const CLSContext = require('zipkin-context-cls');
+const tracer = new Tracer({
+  ctxImpl: new CLSContext('zipkin')
+});
+```
+
+## A not on CLS context vs. explicit context
```
Comment (adriancole) [inline packages/zipkin-context-cls/README.md]: typo "A note"

**Comment by adriancole** [comment]
looks great! request.. could we add a sample of an in-browser trace creation? well.. it might imply we have an http reporter to be realistic (unless scribe works in-browser)

File: packages/zipkin-context-cls/package.json (hunk 1 above)
Comment (eirslett) [inline packages/zipkin-context-cls/package.json]: My fault; it's Openzipkin all over zipkin-js now. We can just do a search-replace later on.

File: packages/zipkin/src/batch-recorder.js (hunk 2 above)
Comment (eirslett) [inline packages/zipkin/src/batch-recorder.js]: Yes, it was actually ported (more or less) directly from Finagle to JavaScript. That's why there's even an option/some/none inplementation in there. We should make zipkin-js more JavaScript-idiomatic later on.

File: packages/zipkin/src/tracer/record.js (hunk 4 above)
Comment (eirslett) [inline packages/zipkin/src/tracer/record.js]: Could be an idea, but I imagine the implementations will diverge somewhat in the longer run.

**Comment by eirslett** [comment]
> looks great! request.. could we add a sample of an in-browser trace creation? well.. it might imply we have an http reporter to be realistic (unless scribe works in-browser)

I'll throw together an isomorphic http transport that can be used both from the browser and from node.js. And a context implementation with zone.js. That will be enough to create an in-browser example. (Strictly speaking, we could probably use ExplicitContext in the browser as well, but overall I imagine zone.js will be prettier)

File: packages/zipkin/src/batch-recorder.js (hunk 3 above)
Comment (eirslett) [inline packages/zipkin/src/batch-recorder.js]: The ServiceName here (as opposed to LocalAddress) is the remote service name. (e.g. "mysql")

File: packages/zipkin/src/tracer/index.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
+require('colors');
```
Comment (sveisvei) [inline packages/zipkin/src/tracer/index.js]: oups, will remove this debugging.

File: packages/zipkin/src/batch-recorder.js (hunk 3 above)
Comment (adriancole) [inline packages/zipkin/src/batch-recorder.js]: ahh so this ends up being the "sa" binary annotation. gotcha

**Pull request merged**